##### ARTI 560 - Computer Vision

## Image Classification with Vision Transformer (ViT) - Exercise

### Objective

In this exercise, you will test the pretrained Vision Transformer (ViT) model on 5 real-world images that you find online.

You will:

1. Download 5 images for different classes in [ImageNet](https://github.com/Waikato/wekaDeeplearning4j/blob/master/docs/user-guide/class-maps/IMAGENET.md).

2. Load the ImageNet class names from a [text file](https://raw.githubusercontent.com/pytorch/hub/master/imagenet_classes.txt).

3. Use ViT to predict the class for each image.

4. Record whether the prediction was correct.

#### Important Note

For this exercise, you MUST use the following KerasHub components:

- [keras_hub.models.ViTImageClassifier](https://keras.io/keras_hub/api/models/vit/vit_image_classifier/)

- [keras_hub.models.ViTImageClassifierPreprocessor](https://keras.io/keras_hub/api/models/vit/vit_image_classifier_preprocessor/)

This ensures your input preprocessing (resizing + normalization) matches what the pretrained ViT model expects.

Do not replace the preprocessor with manual normalization (such as dividing by 255), because it may produce incorrect predictions.

In [10]:
!ls /content

acoustic_guitar.jpg  imagenet_classes.txt  pizza.jpg  sample_data
cat.jpg		     IMAGENET.md	   puppy.jpg  school_bus.jpg


In [11]:
# Import libraries
import tensorflow as tf
import keras_hub
import numpy as np
import os
from PIL import Image

# Load ViTImageClassifierPreprocessor (required by instructor)
preprocessor = keras_hub.models.ViTImageClassifierPreprocessor.from_preset(
    "vit_base_patch16_224_imagenet"
)

# Load ViTImageClassifier with the preprocessor
classifier = keras_hub.models.ViTImageClassifier.from_preset(
    "vit_base_patch16_224_imagenet",
    preprocessor=preprocessor
)

# Path where images are stored in Colab
input_path = "/content"

# ImageNet index mapping for chosen classes
mapping = {
    "cat": 281,
    "puppy": 207,
    "pizza": 963,
    "school_bus": 779,
    "acoustic_guitar": 402
}

# Loop through images and predict
for fname in os.listdir(input_path):

    if fname.endswith(".jpg"):

        image = Image.open(os.path.join(input_path, fname)).convert("RGB")
        img_array = np.array(image)

        # Predict using ViT
        pred_probs = classifier.predict([img_array], verbose=0)
        pred_index = np.argmax(pred_probs)

        true_label = os.path.splitext(fname)[0]

        if true_label in mapping:

            if pred_index == mapping[true_label]:
                print(f"True label: {true_label} (index:{mapping[true_label]}), Predicted Index: {pred_index} --> correct prediction")

            else:
                print(f"True label: {true_label} (index:{mapping[true_label]}), Predicted Index: {pred_index} --> incorrect prediction")

True label: cat (index:281), Predicted Index: 287 --> incorrect prediction
True label: pizza (index:963), Predicted Index: 963 --> correct prediction
True label: puppy (index:207), Predicted Index: 208 --> incorrect prediction
True label: acoustic_guitar (index:402), Predicted Index: 402 --> correct prediction
True label: school_bus (index:779), Predicted Index: 779 --> correct prediction


### Record Your Results
| Image File          | Predicted Label                              | True Label (What you searched) | Correct? (Yes/No) |
| ------------------- | -------------------------------------------- | ------------------------------ | ----------------- |
| cat.jpg             | lynx / similar cat class (index 287)         | cat                            | No                |
| pizza.jpg           | pizza                                        | pizza                          | Yes               |
| puppy.jpg           | Labrador retriever / similar dog (index 208) | puppy                          | No                |
| acoustic_guitar.jpg | acoustic guitar                              | acoustic guitar                | Yes               |
| school_bus.jpg      | school bus                                   | school bus                     | Yes               |
